src/journal/bsod.c

```diff
@@ -0,0 +1,209 @@
+/* SPDX-License-Identifier: LPGL-2.1-or-later */
+
+#include <errno.h>
+#include <getopt.h>
+#include <linux/vt.h>
+#include <sys/ioctl.h>
+#include <stdio.h>
+
+#include "alloc-util.h"
+#include "build.h"
+#include "fd-util.h"
+#include "io-util.h"
+#include "log.h"
+#include "main-func.h"
+#include "sd-id128.h"
+#include "sd-journal.h"
+#include "terminal-util.h"
+
+static void help(void) {
+        printf("%s\n\n"
+               "filters the journal to fetch the first message from the current boot with an emergency log level.\n\n"
+               "   -h --help            Show this help\n"
+               "      --version         Show package version\n",
+               program_invocation_short_name);
+}
+
+static int acquire_first_emergency_log_message(char **ret) {
+        sd_journal *j = NULL;
+        const void *d;
+        size_t l;
+        sd_id128_t boot_id;
+        char boot_id_filter[42];
+        int r;
+
+        r = sd_journal_open(&j, SD_JOURNAL_LOCAL_ONLY);
+        if (r < 0) {
+                log_error_errno(r, "Failed to open journal: %m");
+                goto clean_journal;
+        }
+
+        r = sd_id128_get_boot(&boot_id);
+        if (r < 0) {
+                log_error_errno(r, "Failed to get boot ID: %m");
+                goto clean_journal;
+        }
+
+        snprintf(boot_id_filter, sizeof(boot_id_filter), "_BOOT_ID=" SD_ID128_FORMAT_STR, SD_ID128_FORMAT_VAL(boot_id));
+
+        r = sd_journal_add_match(j, boot_id_filter, 0);
+        if (r < 0) {
+                log_warning_errno(r, "Failed to add boot ID filter: %m");
+                goto clean_journal;
+        }
+
+        r = sd_journal_add_match(j, "_UID=0", 0);
+        if (r < 0) {
+                log_warning_errno(r, "Failed to add User ID filter: %m");
+                goto clean_journal;
+        }
+
+        assert_cc(0 == LOG_EMERG);
+        r = sd_journal_add_match(j, "PRIORITY=0", 0);
+        if (r < 0) {
+                log_warning_errno(r, "Failed to add Emergency filter: %m");
+                goto clean_journal;
+        }
+
+        r = sd_journal_seek_head(j);
+        if (r < 0) {
+                log_warning_errno(r, "Failed to seek to start of jornal: %m");
+                goto clean_journal;
+        }
+
+        r = sd_journal_next(j);
+        if (r < 0) {
+                log_warning_errno(r, "Failed to read next journal entry: %m");
+                goto clean_journal;
+        }
+
```

Comment:
drop this empty line, as we try to keep function calls and their error handling together